### ISSUE TITLE
@discriminator decorator should produce required discriminator property in OpenApi3 emitter

### ISSUE THREAD

**Issue by onexay** [microsoft/typespec#1542]
See following,

```
// Action base
//
@discriminator("name")
model Base {
    @summary("Action type")
    type: CSFI.Models.Common.ActionType;
}

// Non-paramterized action base
//
model NonParameterizedBase<S> extends Base {
    @doc("Action URN")
    name: S;
}

// Parameterized action base
//
model ParameterizedBase<S>  extends Base {
    @doc("Action URN")
    name: S;

    @summary("Parameters")
    @minItems(CSFI.Models.Common.MinActionParameters)
    @maxItems(CSFI.Models.Common.MaxActionParameters)
    parameters: Parameters[];
}
```

Generated OAS3 spec misses, *name* as required property, leading to failure from autorest.

```
     "Models.Actions.Base": {
        "type": "object",
        "properties": {
          "name": {
            "type": "string",
            "description": "Discriminator property for Base."
          },
          "type": {
            "$ref": "#/components/schemas/Models.Common.ActionType"
          }
        },
        "discriminator": {
          "propertyName": "name",
          "mapping": {
              <strings here>
          }
        },
        "required": [
          "type"
        ]
      }
```

```
AutoRest code generation utility [cli version: 3.6.2; node: v19.3.0]
(C) 2018 Microsoft Corporation.
https://aka.ms/autorest
info    | AutoRest core version selected from configuration: ^3.7.0.

There is a new version of AutoRest available (3.6.3).
 > You can install the newer version with with npm install -g autorest@latest

info    |    Loading AutoRest core      'C:\Users\akshayranjan\.autorest\@autorestcore@3.9.3\nodemodules\@autorest\core\dist' (3.9.3)
info    |    Loading AutoRest extension '@autorest/csharp' (latest->3.0.0-beta.20230112.7)
info    |    Loading AutoRest extension '@autorest/modelerfour' (4.25.0->4.25.0)
error   | DiscriminatorNotRequired | Semantic violation: Discriminator must be a required property. (components > schemas > Models.Actions.Base)
  discriminator: {
  propertyName: 'name',
  mapping: {
     <strings> #<referenced schemas>
  }
}
    - file:///C:/Users/akshayranjan/source/repos/CSFI-Service/src/CSFIService/CSFIService.CADL/cadl-output/@cadl-lang/openapi3/openapi.2023-01-01-preview.json:2387:7
fatal   | Process() cancelled due to failure
error   |   Error: Semantic validation failed. There was some errors
error   | Autorest completed with an error. If you think the error message is unclear, or is a bug, please declare an issues at https://github.com/Azure/autorest/issues with the error message you are seeing.
```

Example:
https://cadlplayground.z22.web.core.windows.net/?c=aW1wb3J0ICJAY2FkbC1sYW5nL3Jlc3QiOwoKdXNpbmcgQ2FkbC5IdHRwOwoKZW51bSBOYW1lIHsKICBPcHRpb24xOiAib8YKIizJFjLJFjIiCn0KCi8vIEFjxA8gYmFzZQovLwpAZGlzY3JpbWluYXRvcigibmFtZSIpCm1vZGVsIEJhc8ZkICBAc3VtbWFyeSgix0F0eXDEKcQcxAs6IHN0cmluZzvHZ05vbi1wYXJhbXRlcml6ZWQgYc54xmFOb25QxCVlxybEcTxTPiBleHRlbmRz7QCBZG9jyX1VUk7HfOQAsDogU8h3zU7WdNVx33LVcu8BGMlYc8cp5QD%2FxRHoAR1bXeUAqOYAiOcBuE3FDWlz9gEM5AHmLscqPjvNOjLJOt03Mj47

**Comment by markcowl** [comment]
- emitting template types?
- discriminator should be required

**Comment by markcowl** [comment]
@onexay 

- You are correct that the discriminator should be required in OpenAPI3
- You should use the typespec-autorest emitter (https://www.npmjs.com/package/@azure-tools/typespec-autorest) instead of the OpenAPI3 emitter if you want to run AutoRest against your generated OpenAPI (this emitter supports autorest extensions to OpenAPI2)
- You cannot use a reference to an enum member as the value of a string property (yet).  In your sample you use:
```typespec
model Option1Model is NonParameterizedBase<Name.option1>;
model Option2Model is ParameterizedBase<Name.option2>;
```

For now, you should be able to use a literal string instead:
```typespec
model Option1Model is NonParameterizedBase<"option1">;
model Option2Model is ParameterizedBase<"option2">;
```

Once we have implemented metadata references, you will be able to use:
```typespec
model Option1Model is NonParameterizedBase<Name.option1::value>;
model Option2Model is ParameterizedBase<Name.option2::value>;
```

Here is a playground showing the correct discriminator setup in typespec-autorest

https://cadlplayground.z22.web.core.windows.net/cadl-azure/?c=aW1wb3J0ICJAdHlwZXNwZWMvcmVzdCI7DQoNCnVzaW5nIFR5cGVTcGVjLkh0dHDFGGVudW0gTmFtZSB7DQogIE9wdGlvbjE6ICJvxgoiLMoXMskXMiINCn3EPy8vIEFjxBIgYmFzZcQQDQpAZGlzY3JpbWluYXRvcigibmFtZSIpDQptb2RlbCBCYXPHbCAgQHN1bW1hcnkoIsdF5AC5xCvEHcQMOiBzdHJpbmc7ym9Ob24tcGFyYW10ZXJpemVkIGHwAIDGaE5vblDEJ2XHKMR4PFMgZXh0ZW5kc8dYPskQ7QCXZG9j6QCTVVJO6ACS5ADJOiBT6wCNzWL4AIr%2FAIflAIf%2FAIj4AIjwAUfJa3PILOUBK8US6AFMW13oAMHmAJ%2FnAfNNxQ1pc%2FYBOekCDz47zjgyyTjZNTIiPjs%3D

Which you should be able to use for generation.

Keeping this issue open to track making the discriminator property required in OpenAPI3

**Comment by mikekistler** [comment]
This appears to be fixed as of TypeSpec release 0.45.2. The above playground link does show "name" as required in the generated OpenAPI 3.
```
      "discriminator": "name",
      "required": [
        "name",
        "type"
      ]
```

Can we close this as fixed?

**Comment by onexay** [comment]
Yes.
